keps/sig-network/20201010-kube-proxy-v2/README.md

```diff
@@ -0,0 +1,753 @@
+the network overlay, Cilium to manage app-level security, metallb to provide an external LB for bare-metal clusters, etc).
+
+Along this cambrian explosion of third-party software, the Service object itself received new concepts to improve the
+abstraction, for instance to express topology. This, third-party implementation are expected to update and become more
+complex over time, even if their core doesn't change (ie, the eBPF translation layer is not affected).
+
+This KEP is born from the conviction that more decoupling of the Service object and the actual implementations is required,
+by introducing an intermediate, node-level abstaction provider. This abstraction is expected to be the result of applying
+Kubernetes' `Service` semantics and business logic to a simpler, more stable API.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+- provide a node-level abstraction of the cluster-wide `Service` semantics through an API
+- allow easier, more stable proxy implementations that don't need updates when `Service` business logic changes
+- provide a client library with minimal dependencies
+- include equivalent implementations of in-project ones (userland, iptables and ipvs)
+- (optional) help proxy implementations using the same subsystem (ie iptables) to cooperate more easily
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+
+- provide equivalent implementations of third-party ones
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+Rewrite the kube-proxy to be a "localhost" gRPC API provider that will be accessible via classic TCP (`127.0.0.1:12345`)
+and/or a socket (`unix:///path/to/proxy.sock`).
+
+- it will connect to the API server and watch resources, like the current proxy;
+- then, it will process them, applying Kubernetes specific business logic like topology computation relative to the local host;
+- finally, provide the result of this computation to client via a gRPC watchable API.
+
+The idea is to sent the full state to the client, so implementations don't have to do diff-processing and maintain any internal state.
```

Comment:
```suggestion
The idea is to send the full state to the client, so implementations don't have to do diff-processing and maintain any internal state.
```

Reply:
This assumes the client _wants_ the full state. But the userspace proxy can't do the "overwrite the entire old state with the new state" trick like iptables and ipvs do; it needs diffs. If the daemon only provides full states, then the userspace v2 implementation would need to maintain a copy of the full state and implement diffs itself. So maybe the daemon should support either mode?

Reply:
I wondered the same, but (1) it is my understanding that any k8s controller has a resync loop anyway and (2) could not see any case where the implementation does not have the current state. For instance, with the userspace proxy, the current state will be a list or map of listeners, that can be compared to the expected state, and that is required to apply a diff anyway..